src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -298,4 +359,177 @@ private static int timeUnitToScale(LogicalTypeAnnotation.TimeUnit timeUnit) {
+   * @param type Parquet column type
+   * @param statsMap column stats map to update
+   * @param defaultTimezone default timezone to use for timestamp parsing
+   * @param insertRowsCurrIndex Used for logging the row of index given in insertRows API
+   * @param path dot path of the column
+   * @param isDescendantsOfRepeatingGroup true if the column is a descendant of a repeating group,
+   * @return list of parsed values
+   */
+  private static ParquetBufferValue getGroupValue(
+      Object value,
+      GroupType type,
+      Map<String, RowBufferStats> statsMap,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex,
+      String path,
+      boolean isDescendantsOfRepeatingGroup) {
+    LogicalTypeAnnotation logicalTypeAnnotation = type.getLogicalTypeAnnotation();
+    if (logicalTypeAnnotation == null) {
+      return getStructValue(
+          value,
+          type,
+          statsMap,
+          defaultTimezone,
+          insertRowsCurrIndex,
+          path,
+          isDescendantsOfRepeatingGroup);
+    }
+    if (logicalTypeAnnotation instanceof LogicalTypeAnnotation.ListLogicalTypeAnnotation) {
+      return get3LevelListValue(value, type, statsMap, defaultTimezone, insertRowsCurrIndex, path);
+    }
+    if (logicalTypeAnnotation instanceof LogicalTypeAnnotation.MapLogicalTypeAnnotation) {
+      return get3LevelMapValue(
+          value,
+          type,
+          statsMap,
+          defaultTimezone,
+          insertRowsCurrIndex,
+          path,
+          isDescendantsOfRepeatingGroup);
+    }
+    throw new SFException(
+        ErrorCode.UNKNOWN_DATA_TYPE, logicalTypeAnnotation, type.getClass().getSimpleName());
+  }
+
+  /**
+   * Parses a struct value based on Parquet group logical type. The parsed value is a list of
+   * values, where each element represents a field in the group. For example, an input {@code
+   * {"field1": 1, "field2": 2}} will be parsed as {@code [1, 2]}.
+   */
+  private static ParquetBufferValue getStructValue(
+      Object value,
+      GroupType type,
+      Map<String, RowBufferStats> statsMap,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex,
+      String path,
+      boolean isDescendantsOfRepeatingGroup) {
+    Map<String, ?> structVal =
+        DataValidationUtil.validateAndParseIcebergStruct(path, value, insertRowsCurrIndex);
+    Set<String> extraFields = structVal.keySet();
+    List<Object> listVal = new ArrayList<>(type.getFieldCount());
+    float estimatedParquetSize = 0f;
+    for (int i = 0; i < type.getFieldCount(); i++) {
+      ParquetBufferValue parsedValue =
+          parseColumnValueToParquet(
+              structVal.getOrDefault(type.getFieldName(i), null),
+              type.getType(i),
+              statsMap,
+              defaultTimezone,
+              insertRowsCurrIndex,
+              path,
+              isDescendantsOfRepeatingGroup);
+      extraFields.remove(type.getFieldName(i));
```

Comment:
structVal.keySet() returns a "view" of the underlying map, and any mutations done on extraFields will reflect on the underlying collection too.
From documentation of hashMap.keyset:
```
   * Returns a {@link Set} view of the keys contained in this map.
     * The set is backed by the map, so changes to the map are
     * reflected in the set, and vice-versa.  If the map is modified
     * while an iteration over the set is in progress (except through
     * the iterator's own {@code remove} operation), the results of
     * the iteration are undefined.  The set supports element removal,
     * which removes the corresponding mapping from the map, via the
     * {@code Iterator.remove}, {@code Set.remove},
     * {@code removeAll}, {@code retainAll}, and {@code clear}
     * operations.  It does not support the {@code add} or {@code addAll}
     * operations.
  ```

  Suggest doing `Set<String> extraFields = new HashSet<>(structVal.keySet())` !

Reply:
Fixed in #845, thanks for the fix!